### PR TITLE
[Comb] Narrow Muxes

### PR DESCRIPTION
**Pull request by darthscsi**
Move bit selects into the operands of a mux, so the mux operates on narrow types.

### CODE REVIEW

**Comment by darthscsi** [comment]
Applies O(100) times in a small core test case.

File: lib/Dialect/FIRRTL/FIRRTLFolds.cpp [hunk 1]
```diff
@@ -826,6 +826,36 @@ LogicalResult BitsPrimOp::canonicalize(BitsPrimOp op,
...
     return success();
   }
+  // Mux narrowing
+  // bits(mux(x,a,b)) -> mux(x,bits(a),bits(b))
+  if (auto innerMux = dyn_cast_or_null<MuxPrimOp>(inputOp)) {
+    auto muxTypeWidth = innerMux.getResult()
+                            .getType()
+                            .cast<FIRRTLType>()
+                            .getBitWidthOrSentinel();
+    if (muxTypeWidth > 0 && innerMux->hasOneUse() == 1) {
```
Comment (lattner) [inline lib/Dialect/FIRRTL/FIRRTLFolds.cpp:836]: nit: `hasOneUse` returns a bool, please don't compare against "1".

File: lib/Dialect/FIRRTL/FIRRTLFolds.cpp [hunk 2]
```diff
@@ -826,6 +826,36 @@ LogicalResult BitsPrimOp::canonicalize(BitsPrimOp op,
...
+      // FIRRTL lets constants to muxes be narrower than the mux, which means we
+      // may need to extend the arguments when we move the bitsPrimOp.
+      auto newHi = innerMux.high();
+      if (newHi.getType().cast<FIRRTLType>().getBitWidthOrSentinel() <
+          muxTypeWidth)
```
Comment (lattner) [inline lib/Dialect/FIRRTL/FIRRTLFolds.cpp:841]: This will trigger when the operand has unknown width, which is probably the wrong thing.  I'd recommend wrapping this whole pattern in a check to `hasKnownWidthIntTypesAndNonZeroResult`.

**Review by lattner — APPROVED** [review]
This looks ok to me other than the comments I added.  I'm a bit wary of adding general optimizations to the FIRRTL dialect given how fragile it is with all the sign propagation, implicit extensions, and unknown width issues it has going on. 

Is your interest in cleaning up the generated verilog?  If so, I'd recommend adding this at the comb level instead of the firrtl level.

File: lib/Dialect/FIRRTL/FIRRTLFolds.cpp [hunk 3]
```diff
@@ -826,6 +826,36 @@ LogicalResult BitsPrimOp::canonicalize(BitsPrimOp op,
...
+        newHi = rewriter.createOrFold<PadPrimOp>(innerMux->getLoc(), newHi,
+                                                 muxTypeWidth);
+      newHi = rewriter.createOrFold<BitsPrimOp>(innerMux->getLoc(), newHi,
+                                                op.hi(), op.lo());
+      auto newLow = innerMux.low();
+      if (newLow.getType().cast<FIRRTLType>().getBitWidthOrSentinel() <
+          muxTypeWidth)
+        newLow = rewriter.createOrFold<PadPrimOp>(innerMux->getLoc(), newLow,
```
Comment (lattner) [inline lib/Dialect/FIRRTL/FIRRTLFolds.cpp:849]: Also, is this handling signed inputs to the mux true/false values?

**Review by seldridge — APPROVED** [review]
High level: the optimization makes sense.

I do agree with @lattner that it's probably better to push these into more core dialects. Put differently, I'm fine to aggressively deviate from Scala FIRRTL Compiler optimizations at the core hardware level. I'm reluctant to throw every kitchen sink optimization we can think of into FIRRTL dialect folding.

I still approve, nonetheless. 😄

File: lib/Dialect/Comb/CombFolds.cpp [hunk 4]
```diff
@@ -213,8 +229,9 @@ LogicalResult AndOp::canonicalize(AndOp op, PatternRewriter &rewriter) {
       return success();
     }
 
-    // TODO: Combine multiple constants together even if they aren't at the end.
-    // and(..., c1, c2) -> and(..., c3) where c3 = c1 & c2 -- constant folding
+    // TODO: Combine multiple constants together even if they aren't at the
+    // end. and(..., c1, c2) -> and(..., c3) where c3 = c1 & c2 -- constant
+    // folding
```
Comment (seldridge) [inline lib/Dialect/Comb/CombFolds.cpp:234]: It would be good to back out the unrelated formatting changes.

**Review by seldridge — APPROVED** [review]
lgtm. Nice and simple.

Can you add a test? 

Also, it would be good to back out the unrelated formatting changes here.

**Comment by darthscsi** [comment]
> lgtm. Nice and simple.
> 
> Can you add a test?

Forgot to commit that file.
